pyomo/core/expr/expr_pyomo5.py

```diff
@@ -0,0 +1,3434 @@
+    expression trees.
+    """
+    return expr
+
+
+class clone_counter_context(object):
+    """ Context manager for counting cloning events.
+
+    This context manager counts the number of times that the
+    :func:`clone_expression <pyomo.core.expr.current.clone_expression>`
+    function is executed.
+    """
+
+    _count = 0
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *args):
+        pass
+
+    @property
+    def count(self):
+        """A property that returns the clone count value.
+        """
+        return clone_counter_context._count
+
+#: A clone counter context manager object that simplifies the
+#: use of this context manager.  Specifically, different
+#: instances of this context manger are not necessary.
+clone_counter = clone_counter_context()
+
+
+class mutable_sum_context(object):
+    """ Context manager for mutable sums.
+
+    This context manager is used to compute a sum while
+    treating the summation as a mutable object.
+    """
+
+    def __enter__(self):
+        self.e = _MutableSumExpression([])
+        return self.e
+
+    def __exit__(self, *args):
+        if self.e.__class__ == _MutableSumExpression:
+            self.e.__class__ = SumExpression
+
+#: A context manager object for nonlinear expressions.
+#: This is an instance of the :class:`mutable_sum_contex <pyomo.core.expr.current.mutable_sum_context>` context manager.
+#: Different instances of this context manager are not necessary, unless these context managers are nested.
+nonlinear_expression = mutable_sum_context()
+
+
+class mutable_linear_context(object):
+    """ Context manager for mutable linear sums.
+
+    This context manager is used to compute a linear sum while
+    treating the summation as a mutable object.
+    """
+
+    def __enter__(self):
+        """
+        The :class:`_MutableLinearExpression <pyomo.core.expr.current._MutableLinearExpression>`
+        class is the context that is used to to
+        hold the mutable linear sum.
+        """
+        self.e = _MutableLinearExpression()
+        return self.e
+
+    def __exit__(self, *args):
+        """
+        The context is changed to the
+        :class:`LinearExpression <pyomo.core.expr.current.LinearExpression>`
+        class to transform the context into a nonmutable
+        form.
+        """
+        if self.e.__class__ == _MutableLinearExpression:
+            self.e.__class__ = LinearExpression
+
+#: A context manager object for linear expressions.
+#: This is an instance of the :class:`mutable_linear_contex <pyomo.core.expr.current.mutable_lienar_context>` context manager.
+#: Different instances of this context manager are not necessary unless these context managers are nested.
+linear_expression = mutable_linear_context()
+
+
+#-------------------------------------------------------
+#
+# Visitor Logic
+#
+#-------------------------------------------------------
+
+class SimpleExpressionVisitor(object):
```

Comment:
This file is getting unwieldy.  Should we move the visitor logic into a separate `visitor.py` module?

Reply:
That would make swapping a different expression system more complicated in the future.